crates/ruff_db/src/program.rs

```diff
@@ -0,0 +1,85 @@
+// TODO: Fix clippy warnings in Salsa macros
+#![allow(clippy::needless_lifetimes, clippy::clone_on_copy)]
+
+use crate::{system::SystemPathBuf, Db};
+
+#[salsa::input(singleton)]
+pub struct Program {
+    pub target_version: TargetVersion,
+
+    #[return_ref]
+    pub search_paths: SearchPaths,
```

Comment:
I'd prefer for this field and struct, to be called `search_path_settings: SearchPathSettings`, rather than `search_paths: SearchPaths`. The fields on the struct don't yet represent search paths that we have confidence are valid for module resolution; they're just the raw inputs given to us by the user. The most obvious example of this is the fact that we have a `custom_typeshed: Option<SystemPathBuf>` field. When we're _actually_ resolving modules in the module resolver, we'll always have _a_ path that represents the typeshed stdlib -- but at this stage, it's unknown whether we'll be using a custom (a this point still unvalidated) path given to us by the user to a directory on disk, or `VendoredPath("stdlib")`.

Reply:
First, I don't mind renaming the struct. I just picked a name and probably not a very good one. 

> The fields on the struct don't yet represent search paths that we have confidence are valid for module resolution; they just represent the user's raw settings inputs. 

Can you tell me a bit more about what form of validation is necessary? 

I think we should store the validated configuration on `Program`. The value we store here is **after** reading and resolving the workspace settings. The fact that it is after resolving the setting is important for persistent caching, where we want configurations that resolve in the same settings (e.g., it shouldn't matter if a user explicitly defines a value that is equal to the default or uses the default value) to have equal cache keys.

Reply:
Thanks. I like the name `SearchPathSettings` and I think we can actually defer some of the validation to after creating the program.